### PR TITLE
Make dcdd shock capture optional for heat transfer

### PR DESCRIPTION
**Pull request by mivaia**
### Description

DCDD stabilization was always used for the heat equation. DCDD usage sometimes lead to results that were not expected. More investigation will be needed in the future to understand what is the effect of adding the DCDD stabilization.

DCDD is now disabled by default. A new assembler has been created in order to take it into account. 

### Testing

Modified existing tests using heat transfer. 

### Documentation

Added a set DCDD shock capture option in the multiphysics section. 

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [ ] All in-code documentation related to this PR is up to date (Doxygen format)
- [ ] Lethe documentation is up to date
- [ ] The branch is rebased onto master
- [ ] Code is indented with indent-all and .prm files (examples and tests) with prm-indent
- [ ] If parameters are modified, the tests and the documentation of examples are up to date
- [ ] Changelog (CHANGELOG.md) is up to date if the refactor affects the user experience or the codebase

Pull request related list:
- [ ] No other PR is open related to this refactoring
- [ ] Labels are applied
- [ ] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [ ] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [ ] If any future works is planed, an issue is opened
- [ ] The PR description is cleaned and ready for merge

### CODE REVIEW

File: source/solvers/heat_transfer_assemblers.cc [hunk 1]
```diff
@@ -60,43 +60,46 @@ HeatTransferAssemblerCore<dim>::assemble_matrix(
...
+
+      // // Implementation of a Discontinuity-Capturing Directional Dissipation
+      // // (DCDD) shock capturing scheme. For more information see Tezduyar,
+      // // T. E. (2003). Computation of moving boundaries and interfaces and
+      // // stabilization parameters. International Journal for Numerical
+      // // Methods in Fluids, 43(5), 555-575. Our implementation is based on
+      // // equations (70) and (79), which are adapted for the heat transfer
+      // // solver.
+
+      // // In Tezduyar 2003, this is denoted r in equation (67). The velocity
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:76]: I would delete all these comments. No need to keep it. It's better your new imeplemetaitno.

File: source/solvers/heat_transfer_assemblers.cc [hunk 2]
```diff
@@ -202,42 +206,44 @@ HeatTransferAssemblerCore<dim>::assemble_rhs(
...
+      // // Temperature gradient information for DCDD stabilization
+      // Tensor<1, dim> dcdd_temperature_gradient =
+      //   is_steady(method) ? scratch_data.temperature_gradients[q] :
+      //                       scratch_data.previous_temperature_gradients[0][q];
+
+      // // Implementation of a Discontinuity-Capturing Directional Dissipation
+      // // (DCDD) shock capturing scheme. For more information see Tezduyar,
+      // // T. E. (2003). Computation of moving boundaries and interfaces and
+      // // stabilization parameters. International Journal for Numerical
+      // // Methods in Fluids, 43(5), 555-575. Our implementation is based on
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:218]: Remove all these comments also

File: source/solvers/heat_transfer_assemblers.cc [hunk 3]
```diff
@@ -274,10 +280,10 @@ HeatTransferAssemblerCore<dim>::assemble_rhs(
...
 
           // DCDD shock capturing
-          local_rhs(i) -=
-            rho_cp * nu_dcdd *
-            (scalar_product(grad_phi_T_i, dir_tensor * temperature_gradient)) *
-            JxW;
+          // local_rhs(i) -=
+          //   rho_cp * nu_dcdd *
+          //   (scalar_product(grad_phi_T_i, dir_tensor * temperature_gradient))
+          //   * JxW;
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:286]: To remove

File: source/solvers/heat_transfer_assemblers.cc [hunk 4]
```diff
@@ -742,6 +748,189 @@ HeatTransferAssemblerViscousDissipationVOF<dim>::assemble_rhs(
...
+                      temperature_gradient_unit_vector) -
+        (Utilities::fixed_power<2>(scalar_product(
+           temperature_gradient_unit_vector, velocity_unit_vector)) *
+         outer_product(velocity_unit_vector, velocity_unit_vector));
+
+      // Calculate the artificial viscosity of the shock capture as in equation
+      // (79) of Tezduyar 2003
+      const double nu_dcdd = is_steady(method) ?
+                               0 :
+                               0.5 * h * h * u_mag *
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:821]: I would keep it as a possibility for steady. This way we can debug it maybe more

**Review by blaisb — CHANGES_REQUESTED** [review]
Small things to adress
Good work man. Thanks for this cool refactor
Cna you rebase also?

File: doc/source/parameters/cfd/multiphysics.rst [hunk 5]
```diff
@@ -13,6 +13,7 @@ This subsection defines the multiphysics interface of Lethe and enables the solu
     set heat transfer                   = false
     set viscous dissipation             = false
     set buoyancy force                  = false
+    set DCDD shock capture              = false
```
Comment (blaisb) [inline doc/source/parameters/cfd/multiphysics.rst:16]: I think this parameter would be better in the stabilization section and not the multiphysics sectoin.

**Review by blaisb — APPROVED** [review]
One last tinyc hange, then good to go

File: doc/source/parameters/cfd/multiphysics.rst (hunk 5 above)
Comment (voferreira) [reply doc/source/parameters/cfd/multiphysics.rst:16]: Agreed

File: doc/source/parameters/cfd/multiphysics.rst [hunk 6]
```diff
@@ -40,6 +41,8 @@ This subsection defines the multiphysics interface of Lethe and enables the solu
 
    * ``buoyancy force``: controls if the buoyancy force is taken into account in the Navier-Stokes equations. The buoyancy force is calculated using the Boussinesq approximation.
 
+   * ``DCDD shock capture``: The discontinuity-capturing directional dissipation method is used the stabilize the heat transfer equation. For more information, see Tezduyar, T. E. (2003).
```
Comment (voferreira) [inline doc/source/parameters/cfd/multiphysics.rst:44]:
```suggestion
   * ``DCDD shock capture``: Applies the discontinuity-capturing directional dissipation (DCDD) stabilization term on the heat transfer equation. For more information, see Tezduyar, T. E. (2003).
```

Comment (voferreira) [reply doc/source/parameters/cfd/multiphysics.rst:44]: If you change the section, please don't forget to change the documentation accordingly.

File: include/solvers/heat_transfer_assemblers.h [hunk 7]
```diff
@@ -300,6 +300,46 @@ class HeatTransferAssemblerViscousDissipationVOF
   Parameters::FluidIndicator viscous_dissipative_fluid;
 };
 
+/**
+ * @brief Class that assembles the discontinuity-capturing directional dissipation stablization for the heat transfer
+ * solver. For more information see Tezduyar, T. E. (2003). Computation of
+ * moving boundaries and interfaces and stabilization parameters. International
+ * Journal for Numerical Methods in Fluids, 43(5), 555-575. Our implementation
+ * is based on equations (70) and (79), which are adapted for the heat transfer
+ * solver.
```
Comment (voferreira) [inline include/solvers/heat_transfer_assemblers.h:309]:
```suggestion
 * @brief Class that assembles the discontinuity-capturing directional dissipation stabilization term for the heat transfer
 * solver. For more information see Tezduyar, T. E. (2003). Computation of
 * moving boundaries and interfaces and stabilization parameters. International
 * Journal for Numerical Methods in Fluids, 43(5), 555-575. The implementation
 * is based on equations (70) and (79), which are adapted for the heat transfer
 * solver.
```

File: source/core/parameters_multiphysics.cc [hunk 8]
```diff
@@ -64,7 +64,13 @@ Parameters::Multiphysics::declare_parameters(ParameterHandler &prm)
...
                       "false",
                       Patterns::Bool(),
-                      "Buoyant force calculation <true|false>");
+                      "Buoyant force calculation");
+
+    prm.declare_entry(
+      "DCDD shock capture",
+      "false",
+      Patterns::Bool(),
+      "Use discontinuity-capturing directional dissipation for stabilization in heat transfer <true|false>");
```
Comment (voferreira) [inline source/core/parameters_multiphysics.cc:73]:
```suggestion
      "Apply discontinuity-capturing directional dissipation stabilization term on heat transfer <true|false>");
```

**Review by voferreira — COMMENTED** [review]
Just small comments. I will reread it if you change the parameter section. Otherwise, just pin me and I will approve it. Good job!

**Comment by blaisb** [comment]
@mivaia an advice please rebase master onto your branch instead of merging next time. We always wish to avoid merge commits.

File: source/core/parameters_multiphysics.cc [hunk 9]
```diff
@@ -64,7 +64,7 @@ Parameters::Multiphysics::declare_parameters(ParameterHandler &prm)
     prm.declare_entry("buoyancy force",
                       "false",
                       Patterns::Bool(),
-                      "Buoyant force calculation <true|false>");
+                      "Buoyant force calculation");
```
Comment (blaisb) [inline source/core/parameters_multiphysics.cc:67]:
```suggestion
                      "Buoyant force calculation <true|false>");
```

**Review by blaisb — APPROVED** [review]

**Comment by blaisb** [comment]
Will merge after CI has passed

**Pull request merged**